### ISSUE TITLE
Remove type-checking from the critical path of Angular builds (TypeScript and Template compilation)

### ISSUE THREAD

**Issue by IgorMinar** [angular/angular#43131]
### Description

Developer productivity, velocity, and happiness (no, really!), as well as higher quality of code, and better UX of the final product are directly correlated with how quickly developers can iterate on the code — make a code change and see the result locally.

**The lower the edit & refresh latency (the faster the feedback loop) the better** — this is especially relevant for  interactive workflows (local dev server, unit tests, etc).

**Lower latency means that developers can afford more iterations on solving a problem in a given period of time**, thus enabling them to try various approaches, **be more creative, and to pick the best solution**. If the iteration latency is low, developers often end up shipping code that (barely) meets the functional criteria, and due to a lack of time can't focus on UX or code base quality.

*side note: Latency of committing code and deploying it to production also plays a big role in overall success, but this proposal is specifically focused on the interactive feedback loop.*

Angular has always tried to strike the balance between flexibility that enables creativity and providing guide rails for developers to support them on the path of creating correct, performant, maintainable, accessible, localizable code. One of the means to provide these guide rails has been the use of TypeScript and type-checking to ensure that the code is sound.

Type-checking and the use of type systems are a great way to increase code quality, prevent bugs, and enable automated refactoring of the code base at scale. However, **type-checking also comes at a cost**.

Currently, precious hundreds of milliseconds, but often more commonly many seconds, are being spent on type-checking code of Angular applications as developers edit the code using interactive workflows (like `ng serve`, `ng test --watch`, etc). Not only does this increase the latency, and decrease the number of iterations per minute, it also makes it impossible to see results of a change in the browser if the change doesn't pass the type-checking step. This makes it hard to experiment, and explore, which impacts creativity, and forces developers to either iterate less, or take the time to make the compiler happy, just to find out that they are heading down the wrong direction.

But it doesn't have to be this way... we can have both a faster iteration cycle, as well as all the benefits of the type system and type checking.

### Proposed solution

The current (simplified) processing model of a typical Angular build system is as follows:

```
parse code -> analyze Angular specific metadata (e.g. decorators) -> generate JS code ->
generate TS code needed for type checking (type check blocks) -> ask TypeScript perform type-checking ->
TypeScript to downlevel `.ts` files to `.js` files -> bundle code ->
finally do something with the code (run tests, update dev server, etc)
```

Note that these tasks are performed in a sequence, and type-checking is in the middle of this critical path.

*side note: The Angular CLI currently uses WebPack under the hood, which actually performs type-checking in a separate process, which however still blocks the final steps of the bundling process, resulting in DX similar to (but slightly faster than) fully sequential build pipeline.* 

*side note #2: We can, and do utilize parallelization in order to improve build times, especially when using build systems like Bazel, but even here we achieve parallelization by breaking down bigger apps into smaller compilation units and build those in parallel as much as possible, while still honoring the dependency graph of these compilation units.*

If we removed the type-checking from this critical path we unlock two benefits:

faster builds and rebuilds — type-checking is an expensive operation
enable prototypization and exploration — we'd allow unit tests and dev servers to refresh even if the code doesn't pass all of the type-checks

**The mental model shift proposed by this change is to turn type-checking into a very sophisticated lint check.**

## Implementation details

Much is TBD still and there are many options, ranging from:

- *simple but inefficient:* disable type-checking in the critical path, and use a separate process to run a duplicate TS program that would type-check the code (but not block the dev server refresh / unit test run).
  - We could also support a mode in which the type-checking is disabled during builds, but would still work when running `ng lint` or using [Angular Language Service](https://angular.io/guide/language-service) plugin in IDEs.
- *more evolved, but more efficient*: one possible option would be to embrace [esbuild](https://esbuild.github.io/) as the main build engine with `ngc` (the Angular compiler) as a plugin. In this setup, `ngc` would support a new "type-check-less" mode for efficiently converting Angular metadata and templates into JS code without bothering to analyze and generate code that supports type-checking.
  - Type-checking would still happen out of band, in a way similar to the "simple" solution.
- *extreme*: overhaul many parts of Angular to move away from the "compilation" model, towards "transpilation" (where each `.ts` file can be converted to `.js` file without global program analysis (or global optimization). Then use this transpiler as a highly parallelizable plugin to a build engine like [esbuild](https://esbuild.github.io/). 
  - Type-checking would still happen out of band, in a way similar to the "simple" solution.

## Trade offs

- *developers could write code with bugs* that could be easily caught by type-checking, and waste time debugging the issue in their dev server / unit test. This *could be mitigated with*:
  - IDEs already offer excellent support for low latency type-checking for TypeScript, and with the recent release of Ivy-based Angular Language Service (in v12), we now have the same capability for Angular templates and decorators as well.
  - We can and should still perform type-checking in the background, and **embrace the "eventual consistency" mindset to provide developers with type-checking feedback** in the terminal, and dev server UI (via injected scripts eventually displaying warnings about type issues in the browser)
- *increased complexity of the build pipeline* in order to support parallelization and eventual delivery of the type-checking results to the terminal (for `ng build` and `ng test` or browser (for `ng serve`).
- *duplication of some work* — depending on the implementation we might end up performing a limited amount of work twice, for example source code parsing.
- *increased memory usage of the build pipeline* — similarly to the trade off above, depending on how this is implemented, the build pipeline might end up consuming more memory.

## Web ecosystem compatibility

With the rise of popularity of TypeScript in the web ecosystem, we also observed that a part of the community has embraced the approach of using TypeScript as a glorified linter already. Some examples:

- [babel](https://babeljs.io/) - the TypeScript team @ Microsoft have worked closely to enable syntax-level support (type-stripping) in babel ([blog post](https://devblogs.microsoft.com/typescript/typescript-and-babel-7/))
- [deno](https://deno.land/) - supports TypeScript as the input language without performing type-checking (which is the responsibility of the TypeScript compiler that the developers might choose to run as a "linter").
- [swc](https://swc.rs/) - TypeScript / JavaScript transpiler & bundler (used by deno).
- [esbuild](https://esbuild.github.io/) - TypeScript / JavaScript transpiler & bundler (used by Angular CLI (in a limited way) and [Vite](https://vitejs.dev/))
- [Vite](https://vitejs.dev/) - JavaScript dev server (and build pipeline) with [TypeScript support](https://vitejs.dev/guide/features.html#typescript)
- [bazel](https://bazel.build/) - limited number of recently created projects have started to embrace the use of TypeScript type-checking via [`testonly`](https://docs.bazel.build/versions/main/be/common-definitions.html) targets (according to [aspect.dev](https://www.aspect.dev/))
- [ts-node](https://github.com/TypeStrong/ts-node) - TypeScript execution engine for Node.js with **optional** support for type-checking.

The trend in the community seems to be only accelerating in the direction of using type-checking as an aid, rather than as a hard block in the developer workflows.

### Alternatives considered

Keep on improving the type-checking speed — something we've been focused on for years, and we have reached the point of diminishing returns.

**Comment by Airblader** [comment]
Thanks for the detailed explanation. I'm really happy to see the Angular team making these things a bit more public / transparent.

---

My 2c: I completely understand and welcome the motivation to shorten feedback cycles by separating the type checking process. 
However, often if your types don't check out, your application won't work correctly either. So I would hope that type checking doesn't become an unnoticeable background process that would only be seen in the stdout, but remains (prominently) visible in the browser itself.

**Comment by IgorMinar** [comment]
@Airblader absolutely! the goal is not to make type-checking insignificant - the errors would be very prominent. 

The goal is to give developers an option to ignore the type errors temporarily during the interactive development.

We could even provide an option to make the type-checking errors blocking if developer or team prefers that (controlled via a command line flag, or angular.json config).

**Comment by AwesomeObserver** [comment]
I think esbuild would be good tooling for compilation/translation code pipeline in Angular. 
And moving type check to nonmain process is a chouse.

Also a piece of news, the react framework next.js working on integrating [SWC ](https://swc.rs/)for [optimization pipeline](https://nextjs.org/blog/next-11-1#adopting-rust-based-swc)

**Comment by IgorMinar** [comment]
@AwesomeObserver thanks for the note. It made me realize that I was a bit too prescriptive in the implementation section. We can and should consider using any highly-parallelizable transpiler/bundler, including swc. I updated that section to be less prescriptive.

**Comment by AwesomeObserver** [comment]
@IgorMinar Oh, yes is a good point

**Comment by SanderElias** [comment]
I'm happy with the direction in the current RFC's
Working on large projects, where a 'save & review` cycle can take over 20 seconds, I do feel the pain.
Long term, I would say the 'extreme' option will win out in all scenarios. This would simplify the whole build pipeline, and make scaling so much easier. 
It also perfectly suits the 'atomic' model from ivy, Fits the current es-2020 module model, and would mean you can process all files in parallel when you want. 
Also, this would make Angular less dependent on tooling, and it would become much easier to tailor a build system that would fit the demands of the app. 
It will give the developer more freedom/power on _how_ the end result is reached.
For some apps. putting everything in 1 giant bundle is the optimal solution, while for other apps, going to the extreme of doing any bundling would benefit their use-case. While I don't think either of those extremes will suit any app, I know for sure the place of the line is different for all apps.
Angular devs will be able to make their own trade-offs, that _is_ the thing that counts.

I'm less concerned about memory demands. To a degree, if less memory means slower builds, that is an acceptable trade-off. When less memory means a failing build, that would be a no-go.
If there is a reasonable lower bar, that is ok, as long as that is clearly documented. And properly logged errors do a lot here.
```bash
ng build
`ERROR: this computer doesnt have <amount> of free memory`
```

Back to the other part of this RFC.
Out of band type errors.  
I have no problem here. 
some of my random related thoughts:
- The language server can highlight them, without failing 'ng serve'
- can have a CLI option where they become breaking again?
- should (optionally?) fail `ng build`, but not `ng serve`
- introduce an `ng typecheck` for CI/CD application?
- the above can be an addition to 'ng lint`?

We should recognize there are 2 modes of operating. 
1. develop, 
2. build

### in the development phase we need:
- speed
- speed
- speed
- speed
- optional/ late/eventual consistent type checking 
and don't care about the memory of memory use. or CPU cycles. Most devs have beefy enough machines

### in the build phase we need
- speed
- speed
- quality control
- type checking
And we do care a bit more about memory and CPU cycles. As this will be in CI/CD and most of those systems don't have the same resources as dev machines. (or at least not without a significant cost!)

**Comment by IgorMinar** [comment]
@SanderElias I completely agree with your comment except for one thing - we should remain cognisant of the memory usage for three reasons:

- we want Angular to be accessible to many developers, even those not working on beefy machines. Most of us are lucky to have development machines that usually don't hit the memory ceiling, but developers in other parts of the world, students, and hobbyists don't have that luxury.
- many big projects become so big that building them requires lots of memory, even at Google some of our apps struggle with memory constraints and sometimes fail to build because of OOMs.
- new cloud-enabled or in-browser workflows have tighter memory constraints (for example Stackblitz webcontainers)

So yes, memory usage is a problem to be mindful of, and we can't discount it.

**Comment by flash-me** [comment]
Even though I totally get the mentioned points here, I think, this also should be viewed from the opposing side. 
What about applications that have been build piece by piece and are composed at runtime and therefore don't run into such troubles with long builds and/or `save & review`?

What if people do want type-checking?
Also the build times are hugely affected by external aspects, e.g. 
- Hardware 
- OS (HUUUUUGE difference between *nix and windows)
- Antivirus Software

So there are several options to improve build times. E.g. running tasks in container with a *nix environment is still much faster than running the same task on the windows host.

I would appreciate it, if opting out from type-checking can be turned on/off through a flag and not being hardcoded.

cheers
flash ⚡

**Comment by SanderElias** [comment]
@IgorMinar I think we are in agreement about memory use 😉
self-quote:
> When less memory means a failing build, that would be a no-go.
If there is a reasonable lower bar, that is ok, as long as that is clearly documented. And properly logged errors do a lot here.

I still think that setting a lower bar will help. It is not unreasonable to demand for example 4Gb of free memory. Or 2Gb, and fail instantly with a clean and clear error that stipulates there isn't enough. I would really prefer that over running a process for 2 minutes and then fail with an OOM in the form we get now.

What I really mean with it is, that when memory _is_ available, the CLI can spin up more simultaneous jobs, and give each job more 'memory cache space, making use of the available memory. 
It means that the build-time gap between beefy and under-equipped systems will become even larger. And that is a trade-off I'm willing to live with.

**Comment by IgorMinar** [comment]
@flash-me I agree. I think it was already mentioned earlier that maybe offering 3 way flag would be the best, type checking could then be:

- on and blocking the critical path
- on but out of band
- off (not my personal preference but there might be some workflows where this actually makes sense)

This could be set per dev machine, per workspace, or per project to give you all the flexibility you need.

**Comment by DmitryEfimenko** [comment]
Great discussion and presentation of available options: simple, evolved, extreme. I'm with @SanderElias for the extreme option given the benefits. However, it'd be great to get an idea of the amount of effort involved. I understand "extreme" means the most amount of effort, but what's the very approximate guestimate? I would not want to see this turn into another 2-3-year project.

**Comment by IgorMinar** [comment]
note: the original title of this issue was "Remove type-checking from the critical path of Angular builds (TypeScript and Template compilation)", but through community discussion we found "Out-of-band type-checking for TypeScript and Template compilation" more fitting, so I updated the title.

**Comment by IgorMinar** [comment]
@DmitryEfimenko the extreme option requires changes to several subsystems and more importantly, it represents a mindset shift. It is definitely not cheap. Estimating the effort is hard, especially because we'd need to ensure compatibility with 2500+ apps at Google which we'd be responsible for transitioning to the new pipeline. The good news is that we could break down the project into incrementally deliverable, and on its own valuable, subprojects. A great example of this is the "optional ngmodules" a.k.a "standalone components" project that will present in an RFC in the upcoming days — this project is already heading in the direction of the "extreme" option, and is independently deliverable and very valuable.

If we prioritized this, and not changed our current level of funding and/or support obligations 1.5–3 years for end-to-end execution seems like a good guesstimate for the extreme option.

**Comment by IgorMinar** [comment]
I created a write up on the "extreme" option as a separate proposal: [Transition Angular from the program-based compilation to localized compilation](https://github.com/angular/angular/issues/43165) — please take a look if that option has piqued your interest.

**Comment by bcarson** [comment]
We had a great discussion about this RFC on Angular Nation yesterday, thanks to @IgorMinar, @alxhub @SanderElias, @chaosmonster, @kroeder, @bampakoa and everyone else who joined us! Here's the link for anyone who would like to watch: https://www.youtube.com/watch?v=FbJLC7GxAAs

**Comment by LanderBeeuwsaert** [comment]
As came up also in the Angular Nation discussion, sometimes you just want to quickly try out things and not be bothering with type checking.
In those moments, being able to turn type checking off somehow, would be an interesting (but admittedly potentially 'dangerous') option.
I would imagine that this would be a much less involved change than moving type checking out-of-band.
As such this could be a first quick win and could be used to experiment to what degree removing type-checking delays would help DX.

**Comment by IgorMinar** [comment]
@LanderBeeuwsaert thanks for the note.

Turning off type checking today is not very easy. For starters `tsc` nor `ngc` do not support turning type checking off.

There is an issue open for `tsc` to support [`--transpileOnly` option (microsoft/TypeScript#29651)](https://github.com/microsoft/TypeScript/issues/29651) though. The discussion on that issue is actually similar in many ways to this proposal — for example check out this [comment about how this could speed up CI/CD pipelines](https://github.com/microsoft/TypeScript/issues/29651#issuecomment-598030763).

One thing to keep in mind is that even if `tsc` were to implement this feature, we might still consider using a different tool for the transpilation because `swc` or `esbuild` are usually orders of magnitude faster than `tsc`.

Lastly, I'd much prefer to make type-checking non-blocking rather than push developers towards turning it off, as I do believe that type-checking is really really useful most of the time, I just don't want it to get in the way of fast iteration and exploration.

**Comment by e-oz** [comment]
For me type-checking easily outweighs some milliseconds benefit.  For “10000+ components” apps it might save more time, but in such apps, types are even more important. 

Please at least consider implementing it as a feature that can be turned off.

**Comment by IgorMinar** [comment]
@e-oz thanks for the feedback.

The proposal doesn't include partial type-checking of any kind. The type checking would still happen at the full fidelity and strictness just as it does today. It just proposes to move the type checking to happen out-of-band with respect to the critical edit&refresh path.

Also the cost of type-checking is measured in dozens of seconds for large projects.

Lastly, as discussed above, we should implement this so that the teams/developers can decide if the type-checking should be blocking, non-blocking, or even off.

**Comment by e-oz** [comment]
thanks, @IgorMinar for clarification.
I was just a little bit worried by this 

> The mental model shift proposed by this change is to turn type-checking into a very sophisticated lint check.

and by the plans of using esbuild (because esbuild doesn't check types).

But if it will be totally configurable  - it’s amazing.

**Comment by yharaskrik** [comment]
Love the effort the Angular team is putting into these RFCs, I have been very impressed as of late! (not that I wasn't impressed before that is!).

I agree that the latency really stymies the effectiveness and creativity of developers, especially when making "small" changes (this is a relative term based on your specific scenario obviously). HMR _can_ be a way around this but I have yet to see an implementation that works really well. I have always wanted Angular to embrace a more parallelizable compiler (although the compiler up till now has been fantastic), as large apps start to get a little iffy (other solutions to this obviously, such as incremental compilation from Nx, micro-frontends using module federation etc, but these are all more involved than optimizing the compiler, for the dev obviously, not for the angular term persay). I actually wrote a blog post on the bundle size/speed for our companies code base (https://tech.trellis.org/Angular-v11-vs-v12-Build-Performance/) as I have been interested in this for a while.

I like the idea of being able to allow for type checking to be more of a lint check but would either need to be well thought out since as @e-oz mentioned type checking gets more important the larger the codebase or it would need to be opt in.

We use Nest and Nx at Trellis and the way `nx serve <nest app>` works is it will actually restart the server even if there are type errors and not hold up the refresh cycle. This can be handy, except when it's not, ie. you didn't notice you forgot to pass in a new function parameter and thats why your endpoint doesn't work, for example.

In my opinion **extreme** is the long term solution here, other tools are starting to integrate and support non TS based tooling (like esbuild and swc) as they are written in Rust or other more performant languages. I love Angular and don't want it to fall behind in terms of compilation performance or DX because it is not embracing new tools. Of course I realize this is a major lift and will take a lot of time and effort so I don't expect the Angular team to just go and implement it overnight but do strongly believe we need to keep up to date with tooling. The Angular CLI has always been one of the best selling points of Angular and one of the best built in CLI tools out there, let's not let it fall behind.

All in all, I love this RFC and support optimizing the pipeline wholeheartedly, the better we make it for the developers using it the more people will use it and love it.

**Comment by fetis** [comment]
I double on this. It's especially tedious with unit testing, when a single commented line might bring you a lot of "unused variable, unused import messages". IMO during the development, unless code is compiled (ie syntax is correct) to the JS it should be shipped to the browser. I'm ok to see non-critical errors more as warnings and fix them later when I already have something working

**Comment by matthewjh** [comment]
I would love to see ngc as a plugin that can be used with esbuild and other transpilation tools. If not for our dependency on ngc, we would probably have already switched to "use tsc as a linter and keep it out of the fast dev loop".

> I double on this. It's especially tedious with unit testing, when a single commented line might bring you a lot of "unused variable, unused import messages". IMO during the development, unless code is compiled (ie syntax is correct) to the JS it should be shipped to the browser. I'm ok to see non-critical errors more as warnings and fix them later when I already have something working.
> 
> My current turnover on a project is between 30 and 60 secs, you loose focus and context on a such long time span.

You can turn off `noUnusedLocals`,  `noUnusedParameters` etc. in your tsconfig:

https://www.typescriptlang.org/tsconfig#noUnusedLocals

**Comment by jpike88** [comment]
Angular's core tools have always been greatly lacking in development cycle speed, helping give rise to a whole ecosystem of bundlers that just do that part better. The demand for performant development has always been there, it shaves precious hours off everyone's time. The last few major versions of Angular have done little to help with this. Throwing more stuff on the pile needs to stop and performance needs to take the front seat. Hopefully the upper management will reconsider what the priorities should be, I don't understand how such a slow cycle was ever acceptable.

**Comment by jpike88** [comment]
Relevant: 

https://news.ycombinator.com/item?id=30074414

https://kdy1.dev/posts/2022/1/tsc-go
